### PR TITLE
Upgrade Vite to latest

### PR DESCRIPTION
**Pull request by natemoo-re**
## Changes

- Another attempt at https://github.com/withastro/astro/pull/1866 (that branch had gotten unwieldy.)
- Un-vendors Vite, upgrades to `vite@2.8.x-beta` as there are some fixes there that haven't landed in `vite@2.7.x`

## Testing

Tracking tests in CI

## Docs

N/A

### CODE REVIEW

**Comment by matthewp** [comment]
Blocked by https://github.com/vitejs/vite/issues/6576

**Comment by matthewp** [comment]
This is blocked by another Vite bug that I'm submitting a PR for shortly.

**Comment by matthewp** [comment]
Blocked by https://github.com/vitejs/vite/pull/6589

**Comment by natemoo-re** [comment]
Looks like the final issues caused by Svelte are now passing! Just two issues left.

**Comment by natemoo-re** [comment]
This is super close, one last absolute path issue on Windows.

File: examples/with-nanostores/src/pages/index.astro [hunk 1]
```diff
@@ -3,7 +3,6 @@
 import AdminsReact from '../components/AdminsReact.jsx';
 import AdminsSvelte from '../components/AdminsSvelte.svelte';
 import AdminsVue from '../components/AdminsVue.vue';
-import AdminsPreact from '../components/AdminsPreact.jsx';
```
Comment (natemoo-re) [inline examples/with-nanostores/src/pages/index.astro:6]: This is a change worth calling out... Had to drop the `@nanostores/preact` usage from this example to get everything passing. I think there's some issue with deduping going on, but I'm hoping it's just in our monorepo. Will be sure to check this once we cut a release.

File: packages/astro/package.json [hunk 2]
```diff
@@ -106,7 +105,7 @@
     "strip-ansi": "^7.0.1",
     "supports-esm": "^1.0.0",
     "tsconfig-resolver": "^3.0.1",
-    "vite": "~2.6.10",
+    "vite": "^2.8.0-beta.7",
```
Comment (natemoo-re) [inline packages/astro/package.json:108]: No more vendored `vite`! It's a real dependency.

File: packages/astro/src/core/create-vite.ts [hunk 3]
```diff
@@ -16,18 +16,15 @@ import { resolveDependency } from './util.js';
...
 	'node-fetch',
 	'prismjs',
 	'shiki',
 	'shorthash',
 	'unified',
 	'whatwg-url',
 ]);
-const ALWAYS_NOEXTERNAL = new Set([
-	'astro', // This is only because Vite's native ESM doesn't resolve "exports" correctly.
-]);
```
Comment (natemoo-re) [inline packages/astro/src/core/create-vite.ts:30]: This comment is no longer true, so `astro` can now be optimized by Vite.

File: packages/astro/src/runtime/server/hydration.ts [hunk 4]
```diff
@@ -1,34 +1,12 @@
 import type { AstroComponentMetadata } from '../../@types/astro';
 import type { SSRElement, SSRResult } from '../../@types/astro';
-import { valueToEstree } from 'estree-util-value-to-estree';
-import * as astring from 'astring';
```
Comment (natemoo-re) [inline packages/astro/src/runtime/server/hydration.ts:4]: These packages we needlessly complex and weren't being handled well by Vite.

File: packages/astro/src/runtime/server/hydration.ts [hunk 5]
```diff
@@ -1,34 +1,12 @@
...
-		if (node.raw != null) {
-			// escape closing script tags in strings so browsers wouldn't interpret them as
-			// closing the actual end tag in HTML
-			state.write(node.raw.replace('</script>', '<\\/script>'));
-		} else {
-			GENERATOR.Literal(node, state);
-		}
-	},
-};
+import serializeJavaScript from 'serialize-javascript';
```
Comment (natemoo-re) [inline packages/astro/src/runtime/server/hydration.ts:4]: This was replaced by `serializeJavaScript` which handles everything for us.

File: packages/astro/src/vite-plugin-astro/index.ts [hunk 6]
```diff
@@ -64,15 +64,20 @@ export default function astro({ config, logging }: AstroPluginOptions): vite.Plu
 			let { filename, query } = parseAstroRequest(id);
 			if (query.astro) {
 				if (query.type === 'style') {
+					if (filename.startsWith('/@fs')) {
+						filename = filename.slice('/@fs'.length);
+					} else if (filename.startsWith('/') && !ancestor(filename, config.projectRoot.pathname)) {
+						filename = new URL('.' + filename, config.projectRoot).pathname;
+					}
+
```
Comment (natemoo-re) [inline packages/astro/src/vite-plugin-astro/index.ts:72]: @matthewp any idea if this was old code that snuck back in via a rebase?

File: examples/with-nanostores/src/pages/index.astro (hunk 1 above)
Comment (FredKSchott) [reply examples/with-nanostores/src/pages/index.astro:6]: Thanks for the callout. If you're worried about not being able to resolve this before end of week, I'd suggest creating an issue to track adding `@nanostores/preact` back.

**Review by FredKSchott — APPROVED** [review]
LGTM! What a huge effort @natemoo-re !

Just a callout that this will be released as `npm install astro@next` before it goes out to everyone, so there will be time to test with real-user data post-merge.

**Review by jonathantneal — APPROVED** [review]
For what this change is, it was surprising clear. Well done, @natemoo-re!

**Pull request merged**